tests/net/tcp/src/main.c

```diff
@@ -1118,6 +1128,269 @@ ZTEST(net_tcp, test_client_fin_wait_2_ipv4)
+		device_initial_seq = ntohl(th->th_seq);
+		seq = 0U;
+		ack = ntohl(th->th_seq) + 1U;
+		reply = prepare_syn_ack_packet(af, htons(MY_PORT),
+					       th->th_sport);
+		seq++;
+		t_state = T_SYN_ACK;
+		break;
+	case T_SYN_ACK:
+		test_verify_flags(th, ACK);
+		/* connection is success */
+		reply = prepare_data_packet(af, htons(MY_PORT),
+					    th->th_sport, "A", 1U);
+		t_state = T_DATA_ACK;
+		break;
+	case T_DATA_ACK:
+		test_verify_flags(th, ACK);
+		test_sem_give();
+		reply = NULL;
+		t_state = T_FIN;
+		return;
+	case T_FIN:
+		test_verify_flags(th, FIN | ACK);
+		zassert_true(get_rel_seq(th) == 1,
+			     "%s:%d unexpected sequence number in original FIN, got %d",
+			     __func__, __LINE__, get_rel_seq(th));
+		zassert_true(ntohl(th->th_ack) == 2,
+			     "%s:%d unexpected acknowlegdement in original FIN, got %d",
+			     __func__, __LINE__, ntohl(th->th_ack));
+		t_state = T_FIN_1;
+		/* retransmit the data that we already send*/
+		reply = prepare_data_packet(af, htons(MY_PORT),
+					    th->th_sport, "A", 1U);
+		seq++;
+		break;
+	case T_FIN_1:
+		test_verify_flags(th, FIN | ACK);
+		/* retransmitted FIN should have the same sequence number*/
+		zassert_true(get_rel_seq(th) == 1,
+			     "%s:%i unexpected sequence number in retransmitted FIN, got %d",
+			     __func__, __LINE__, get_rel_seq(th));
+		zassert_true(ntohl(th->th_ack) == 2,
+			     "%s:%i unexpected acknowlegdement in retransmitted FIN, got %d",
+			     __func__, __LINE__, ntohl(th->th_ack));
+		ack = ack + 1U;
+		t_state = T_FIN_2;
+		reply = prepare_ack_packet(af, htons(MY_PORT), th->th_sport);
+		break;
+	case T_FIN_2:
+		t_state = T_FIN_ACK;
+		reply = prepare_fin_packet(af, htons(MY_PORT), th->th_sport);
+		break;
+	case T_FIN_ACK:
+		test_verify_flags(th, ACK);
+		test_sem_give();
+		return;
+	default:
+		zassert_true(false, "%s unexpected state", __func__);
+		return;
+	}
+
+	ret = net_recv_data(iface, reply);
+	if (ret < 0) {
+		goto fail;
+	}
+
+	if (t_state == T_FIN_2) {
+		goto send_next;
+	}
+
+	return;
+fail:
+	zassert_true(false, "%s failed", __func__);
+}
+
+/* Test case scenario IPv4
+ *   expect SYN,
+ *   send SYN ACK,
+ *   expect ACK,
+ *   expect Data,
+ *   send ACK,
+ *   expect FIN,
+ *   resend Data,
+ *   expect FIN,
+ *   send ACK,
+ *   send FIN,
+ *   expect ACK
+ *   any failures cause test case to fail.
+ */
+ZTEST(net_tcp, test_client_fin_wait_1_retransmit_ipv4)
+{
+	struct net_context *ctx;
+	int ret;
+
+	t_state = T_SYN;
+	test_case_no = 10;
+	seq = ack = 0;
+
+	ret = net_context_get(AF_INET, SOCK_STREAM, IPPROTO_TCP, &ctx);
+	if (ret < 0) {
+		zassert_true(false, "Failed to get net_context");
+	}
+
+	net_context_ref(ctx);
+
+	ret = net_context_connect(ctx, (struct sockaddr *)&peer_addr_s,
+				  sizeof(struct sockaddr_in),
+				  NULL,
+				  K_MSEC(100), NULL);
+	if (ret < 0) {
+		zassert_true(false, "Failed to connect to peer");
+	}
+
+	ret = net_context_recv(ctx, test_tcp_recv_cb, K_MSEC(200), NULL);
```

Comment:
@rlubos: While rebasing I found that the latest commit of this PR breaks. This context_recv never completes. It is the only place in these tests where the recv is called from a client context. Might it have something to do with the recently merged PR: https://github.com/zephyrproject-rtos/zephyr/pull/61336

Reply:
I've debuged this a bit and it seems like a race condition in the test. The DATA packet is received by the TCP stack **before** the test actually installs recv callback with `net_context_recv()`, so it gets dropped:
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/ip/tcp.c#L982

The change from https://github.com/zephyrproject-rtos/zephyr/pull/61336 merely revealed that the `net_context_recv()` times out, as before that the timeout result was not returned correctly.

Reply:
Calling `net_context_recv()` before connect, with K_NO_WAIT timeout (just to install the recv cb) seems to help. The semaphore takes care of checking whether the data actually arrived.

Reply:
Thanks for your input. Finally I removed the `net_context_recv()` altogether as is not required to test the network stack.